kafka/consumer-group.go

```diff
@@ -0,0 +1,81 @@
+package kafka
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/Shopify/sarama"
+	"github.com/bsm/sarama-cluster"
+)
+
+var tick = time.Millisecond * 4000
+
+type ConsumerGroup struct {
+	Consumer *cluster.Consumer
+	Incoming chan Message
+	Closer   chan bool
+	Errors   chan error
+}
+
+type Message struct {
+	message  *sarama.ConsumerMessage
+	consumer *cluster.Consumer
+}
+
+func (M Message) GetData() []byte {
+	return M.message.Value
+}
+
+func (M Message) Commit() {
+	M.consumer.MarkOffset(M.message, "metadata")
+	//M.consumer.CommitOffsets()
+	//log.Printf("Offset : %d, Partition : %d", M.message.Offset, M.message.Partition)
+}
+
+func NewConsumerGroup(brokers []string, topic string, group string, offset int64) (*ConsumerGroup, error) {
+	config := cluster.NewConfig()
+	config.Group.Return.Notifications = true
+	config.Consumer.Return.Errors = true
+	config.Consumer.MaxWaitTime = 50 * time.Millisecond
+	config.Consumer.Offsets.Initial = offset
+
+	consumer, err := cluster.NewConsumer(brokers, group, []string{topic}, config)
+	if err != nil {
+		return nil, fmt.Errorf("Bad NewConsumer of %q: %s", topic, err)
+	}
+
+	cg := ConsumerGroup{
+		Consumer: consumer,
+		Incoming: make(chan Message),
+		Closer:   make(chan bool),
+		Errors:   make(chan error),
+	}
+
+	go func() {
+		defer cg.Consumer.Close()
+		log.Info(fmt.Sprintf("Started kafka consumer of topic %q group %q", topic, group), nil)
+		for {
+			select {
+			case err := <-cg.Consumer.Errors():
+				log.Error(err, nil)
+				cg.Errors <- err
+			default:
+				select {
+				case msg := <-cg.Consumer.Messages():
+					cg.Incoming <- Message{msg, cg.Consumer}
+				case n, more := <-cg.Consumer.Notifications():
+					if more {
+						log.Trace("Rebalancing group", log.Data{"topic": topic, "group": group, "partitions": n.Current[topic]})
```

Comment:
I agree that a library should _generally_ not force logging decisions on you.

The kafka package was created to capture a common usage pattern across the pipeline's (and now also perceived `go-ns` users') microservices, and hence replicated that user need.

Given that it was meant to be a thin abstraction layer, and that you've pointed out (verbally) that it doesn't really add much value in general, this is perhaps one place where value is added (albeit using a forced/controversial design decision!) by taking care of the monitoring and logging of re-partitioning.

I guess I'm saying that if you want to avoid that "added" value, then don't use this library. But you're already using `go-ns` so use our logging, too.